cylc/flow/network/graphql.py

```diff
@@ -0,0 +1,153 @@
+        schema,  # type: GraphQLSchema
+        document_ast,  # type: Document
+        *args,  # type: Any
+        **kwargs  # type: Any
+):
+    # type: (...) -> Union[ExecutionResult, Observable]
+    """Validate schema, and execute request doc against it."""
+    do_validation = kwargs.get("validate", True)
+    if do_validation:
+        validation_errors = validate(schema, document_ast)
+        if validation_errors:
+            return ExecutionResult(errors=validation_errors, invalid=True)
+
+    result = execute(schema, document_ast, *args, **kwargs)
+
+    if kwargs.get('strip_null', False):
+        if kwargs.get('return_promise', False):
+            return result.then(null_stripper)
+        return null_stripper(result)
+    return result
+
+
+class GraphQLCoreBackend(GraphQLBackend):
+    """GraphQLCoreBackend will return a document using the default
+    graphql executor"""
+
+    def __init__(self, executor=None):
+        # type: (Optional[Any]) -> None
+        self.execute_params = {"executor": executor}
+
+    def document_from_string(self, schema, document_string):
+        # type: (GraphQLSchema, Union[Document, str]) -> GraphQLDocument
+        """Parse string and setup request docutment for execution."""
+        if isinstance(document_string, ast.Document):
+            document_ast = document_string
+            document_string = print_ast(document_ast)
+        else:
+            if not isinstance(document_string, str):
+                logger.error("The query must be a string")
+            document_ast = parse(document_string)
+        return GraphQLDocument(
+            schema=schema,
+            document_string=document_string,
+            document_ast=document_ast,
+            execute=partial(
+                execute_and_validate,
+                schema,
+                document_ast,
+                **self.execute_params
+            ),
+        )
+
+
+# -- Middleware --
+
+class IgnoreFieldMiddleware:
+    """Set to null/None type undesired field values for stripping."""
+
+    ALLOW_TYPES = [0, 0., False]
+
+    def resolve(self, next, root, info, **args):
+        """Middleware resolver; handles field according to operation."""
+        if info.operation.operation == 'query':
+            return self.async_resolve(next, root, info, **args)
+        if info.operation.operation == 'subscription':
+            return self.null_setter(next(root, info, **args))
+        return next(root, info, **args)
+
+    async def async_resolve(self, next, root, info, **args):
+        """Set type to null after awaited result if empty/null-like."""
+        result = await next(root, info, **args)
+        return self.null_setter(result)
+
+    def null_setter(self, result):
+        """Set type to null if result is empty/null-like."""
+        # If result is empty... could be some other condition.
+        # excluded False, as could be a flag turned off.
+        if result or result in self.ALLOW_TYPES:
```

Comment:
This is testing whether the result not empty or an allowed empty type (i.e. `False`).

Reply:
BTW - the suggested change would miss `[]`, `{}`, and `""`, which we also want to filter out... so the other way to test it would be:
```
EMPTY_TYPES = ([], {}, "")
if result in EMPTY_TYPES:
    return None
return result
```
Any preference? (picked the original to avoid checking on each empty type for all results, although maybe `if result` does this under the hood)

Reply:
If I understand the null-stripping correctly a limitation of this approach is that we cant "reset" fields. So for example if a field is set to `'foo'`, but then we want to reset it to `None` the change won't be reflected in the deltas. This creates potential for data stores synced via GraphQL deltas to get subtly out of sync.

Taking a look through the PB schema I don't think this is likely to be a major issue, the best example I can find is `latest_message` which is a field of the task proxy but which we might expect to get reset if you were to manually reset the task state to waiting. Not the most compelling example but illustrative of the general issue.

Options:

1. Document this limitation:

    * Might be enough just to document that one sholdn't reset fields on the data store to `None`.
    * Kinda like saying the data store is "progressive" i.e. you can add or increment but you cant take away, which might actually be a good data model for a workflow engine.

2. Add an unset type:

    * Create a constant for unset fields (would need one constant per data type).
    * E.G. an integer field might have an unset value of `-99999` and a string field might have an unset value of `__UNSET__`.
    * Use this value to determine which fields to strip.

3. Something else?

I'm happy with whatever solution, option (1) might be sufficient if it turns out that we aren't presently resetting any fields.

Reply:
I think what I have in place is in line with (1), with some caveats:
- The actual data-store (protobuf) can only have values of the same type (so None is not valid for a string type, it would be `'foo'` or `''`.
- Also (like mentioned in riot), if there is some meaningful change to a field then it can be reflected in the "reset" value (i.e. things like `latest_message` actually set to a non-empty (`"pending"`?) when a significant change requires a "reset")

Open to ideas, put if we can't treat all empty types (i.e. `""`) the same for all fields, then we'd need to only strip (set null) according to the field..